src/plugins/janus_audiobridge.c

```diff
@@ -8655,29 +8658,55 @@ static void *janus_audiobridge_participant_thread(void *data) {
 						janus_audiobridge_buffer_packet_destroy(bpkt);
 						break;
 					}
+
 					rtp = (janus_rtp_header *)bpkt->buffer;
 					/* If this is Opus, check if there's a packet gap we should fix with FEC */
 					use_fec = FALSE;
 					if(!first && participant->codec == JANUS_AUDIOCODEC_OPUS && participant->fec) {
-						if(ntohs(rtp->seq_number) == (participant->expected_seq + 1)) {
-							/* Lost a packet here? Use FEC to recover */
+						if(ntohs(rtp->seq_number) != (participant->expected_seq)) {
+							lost_packets = ntohs(rtp->seq_number) - participant->expected_seq;
+							JANUS_LOG(LOG_ERR, "[%d] got rtp seq, expected_seq: [%d], diff: [%d]\n", ntohs(rtp->seq_number), participant->expected_seq, lost_packets);
+
+							int32_t output_samples;
+							opus_decoder_ctl(participant->decoder, OPUS_GET_LAST_PACKET_DURATION(&output_samples));
+
+							for(int i=1; i < lost_packets; i++) {
+								pkt = g_malloc(sizeof(janus_audiobridge_rtp_relay_packet));
+								pkt->data = g_malloc0(OPUS_SAMPLES * (participant->stereo ? 2 : 1) * sizeof(opus_int16));
+								pkt->ssrc = 0;
+								pkt->timestamp = participant->last_timestamp + 960 * i;
+								pkt->seq_number = participant->last_seq + 1 * i;
+								/* This is a redundant packet, so we can't parse any extension info */
+								pkt->silence = FALSE;
+								pkt->length = opus_decode(participant->decoder, NULL, 0, (opus_int16 *)pkt->data, output_samples, 0);
+								JANUS_LOG(LOG_ERR, "[%d] packet concealed, length: [%d], timestamp: [%d] stereo: [%d]\n", pkt->seq_number, pkt->length, pkt->timestamp, participant->stereo ? 2 : 1);
+								janus_mutex_lock(&participant->qmutex);
+								participant->inbuf = g_list_append(participant->inbuf, pkt);
+								janus_mutex_unlock(&participant->qmutex);
+							}
+
 							use_fec = TRUE;
 						}
 					}
 					first = FALSE;
 					if(use_fec) {
 						/* There was a gap, try to get decode from redundant info first */
 						pkt = g_malloc(sizeof(janus_audiobridge_rtp_relay_packet));
-						pkt->data = g_malloc0(BUFFER_SAMPLES*sizeof(opus_int16));
+						pkt->data = g_malloc0(OPUS_SAMPLES * (participant->stereo ? 2 : 1) * sizeof(opus_int16));
 						pkt->ssrc = 0;
-						pkt->timestamp = participant->last_timestamp + 960;	/* FIXME */
-						pkt->seq_number = participant->expected_seq;		/* FIXME */
+						pkt->timestamp = ntohl(rtp->timestamp) - 960;	/* FIXME */
+						pkt->seq_number = ntohs(rtp->seq_number) - 1;		/* FIXME */
 						/* This is a redundant packet, so we can't parse any extension info */
 						pkt->silence = FALSE;
-						/* Decode the lost packet using fec=1 */
+
```

Comment:
I'm not sure why you completely changed the way we handle FEC? The existing FEC management should remain the same, if not broken. See comments above.

Reply:
FEC management which were implemented earlier didn't work in case of gap > 1, I didn't change too much, just adapted it to work using latest arrived rtp packet info instead of expected value based on last decoded packet.